File: crm/app/internal/adapters/inbound/http/dedup_handler.go
// Package http — dedup HTTP handlers.
// Interface-based; no direct import of sqlite or application packages.
//
// Auth: DEFERRED — endpoints are LAN-trust only (per plan). No auth middleware added here.
package http

import (
	"context"
	"encoding/json"
	"log"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/vantt/data-integration/crm/app/internal/domain"
)

// DedupQuerier is the minimal read interface the dedup list handler needs.
type DedupQuerier interface {
	ListCandidates(ctx context.Context, status string) ([]domain.DedupCandidate, error)
	GetCandidate(ctx context.Context, candidateID string) (*domain.DedupCandidate, error)
	UpdateCandidateStatus(ctx context.Context, candidateID, status string, reviewedBy *string) error
}

// DedupMerger is the minimal merge interface the merge handler needs.
type DedupMerger interface {
	Merge(ctx context.Context, survivingPartyID, mergedPartyID, reason string, mergedBy *string) (string, error)
}

// DedupHandler holds dependencies for all dedup HTTP handlers.
type DedupHandler struct {
	querier DedupQuerier
	merger  DedupMerger
}

// NewDedupHandler constructs a DedupHandler.
func NewDedupHandler(q DedupQuerier, m DedupMerger) *DedupHandler {
	return &DedupHandler{querier: q, merger: m}
}

// RegisterRoutes wires dedup endpoints onto a chi router.
func (h *DedupHandler) RegisterRoutes(r chi.Router) {
	r.Get("/api/dedup/candidates", h.listCandidates)
	r.Post("/api/dedup/merge", h.mergeCandidates)
	r.Post("/api/dedup/candidates/{id}/reject", h.rejectCandidate)
}

// listCandidates handles GET /api/dedup/candidates
// Optional query param: ?status=pending|merged|rejected (default: pending)
func (h *DedupHandler) listCandidates(w http.ResponseWriter, r *http.Request) {
	status := r.URL.Query().Get("status")
	if status == "" {
		status = "pending"
	}

	candidates, err := h.querier.ListCandidates(r.Context(), status)
	if err != nil {
		log.Printf("dedup: list candidates: %v", err)
		writeError(w, http.StatusInternalServerError, "failed to list candidates")
		return
	}
	if candidates == nil {
		candidates = []domain.DedupCandidate{}
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"status":     "ok",
		"count":      len(candidates),
		"candidates": candidates,
	})
}

type mergeRequest struct {
	SurvivingID string  `json:"surviving_id"`
	MergedID    string  `json:"merged_id"`
	Reason      string  `json:"reason"`
	MergedBy    *string `json:"merged_by,omitempty"`
}

// mergeCandidates handles POST /api/dedup/merge
func (h *DedupHandler) mergeCandidates(w http.ResponseWriter, r *http.Request) {
	var req mergeRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.SurvivingID == "" || req.MergedID == "" {
		writeError(w, http.StatusBadRequest, "surviving_id and merged_id are required")
		return
	}

	mergeID, err := h.merger.Merge(r.Context(), req.SurvivingID, req.MergedID, req.Reason, req.MergedBy)
	if err != nil {
		log.Printf("dedup: merge %s→%s: %v", req.MergedID, req.SurvivingID, err)
		writeError(w, http.StatusInternalServerError, "merge failed")
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"status":   "ok",
		"merge_id": mergeID,
	})
}

// rejectCandidate handles POST /api/dedup/candidates/{id}/reject
func (h *DedupHandler) rejectCandidate(w http.ResponseWriter, r *http.Request) {
	candidateID := chi.URLParam(r, "id")
	if candidateID == "" {
		writeError(w, http.StatusBadRequest, "candidate id required")
		return
	}

	c, err := h.querier.GetCandidate(r.Context(), candidateID)
	if err != nil {
		log.Printf("dedup: get candidate %s: %v", candidateID, err)
		writeError(w, http.StatusInternalServerError, "operation failed")
		return
	}
	if c == nil {
		writeError(w, http.StatusNotFound, "candidate not found")
		return
	}

	if err := h.querier.UpdateCandidateStatus(r.Context(), candidateID, "rejected", nil); err != nil {
		log.Printf("dedup: reject candidate %s: %v", candidateID, err)
		writeError(w, http.StatusInternalServerError, "operation failed")
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{"status": "ok", "candidate_id": candidateID})
}

// writeJSON encodes v as JSON with the given status code.
func writeJSON(w http.ResponseWriter, code int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	_ = json.NewEncoder(w).Encode(v)
}

// writeError writes a JSON error response.
func writeError(w http.ResponseWriter, code int, msg string) {
	writeJSON(w, code, map[string]string{"error": msg})
}

File: crm/python/adapters/inbound/http/dedup_handler.py
"""FastAPI router — dedup HTTP handlers.

Auth: DEFERRED — LAN-trust only, no auth middleware.
Routes mirror dedup_handler.go (chi → FastAPI).
"""
from __future__ import annotations

import logging
from typing import Any, Optional, Protocol

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from crm.python.domain.entities.party import DedupCandidate

log = logging.getLogger(__name__)

router = APIRouter(prefix="/api")


# ── Service protocols ─────────────────────────────────────────────────────────

class DedupQuerier(Protocol):
    def list_candidates(self, status: str) -> list[DedupCandidate]: ...
    def get_candidate(self, candidate_id: str) -> Optional[DedupCandidate]: ...
    def update_candidate_status(
        self, candidate_id: str, status: str, reviewed_by: Optional[str]
    ) -> None: ...


class DedupMerger(Protocol):
    def merge(
        self,
        surviving_party_id: str,
        merged_party_id: str,
        reason: str,
        merged_by: Optional[str],
    ) -> str: ...

    def undo_merge(self, merge_id: str, undone_by: Optional[str] = None) -> None: ...


# ── Pydantic request models ───────────────────────────────────────────────────

class MergeRequest(BaseModel):
    surviving_id: str
    merged_id: str
    reason: str = ""
    merged_by: Optional[str] = None


class UndoMergeRequest(BaseModel):
    undone_by: Optional[str] = None


# ── Handler factory ───────────────────────────────────────────────────────────

def make_dedup_router(querier: DedupQuerier, merger: DedupMerger) -> APIRouter:
    """Return a configured APIRouter with all dedup endpoints wired."""

    # GET /api/dedup/candidates?status=pending|merged|rejected
    @router.get("/dedup/candidates")
    def list_candidates(status: str = "pending") -> Any:
        candidates = querier.list_candidates(status)
        return {
            "status": "ok",
            "count": len(candidates),
            "candidates": candidates or [],
        }

    # POST /api/dedup/merge/{candidate_id}
    @router.post("/dedup/merge/{candidate_id}")
    def merge_candidate(candidate_id: str, body: MergeRequest) -> Any:
        if not body.surviving_id or not body.merged_id:
            raise HTTPException(
                status_code=400,
                detail="surviving_id and merged_id are required",
            )
        candidate = querier.get_candidate(candidate_id)
        if candidate is None:
            raise HTTPException(status_code=404, detail="candidate not found")
        try:
            merge_id = merger.merge(
                body.surviving_id,
                body.merged_id,
                body.reason,
                body.merged_by,
            )
        except ValueError as exc:
            raise HTTPException(status_code=422, detail=str(exc)) from exc
        querier.update_candidate_status(candidate_id, "merged", body.merged_by)
        return {"status": "ok", "merge_id": merge_id, "candidate_id": candidate_id}

    # POST /api/dedup/undo/{merge_id}
    @router.post("/dedup/undo/{merge_id}")
    def undo_merge(merge_id: str, body: UndoMergeRequest = UndoMergeRequest()) -> Any:
        try:
            merger.undo_merge(merge_id, body.undone_by)
        except ValueError as exc:
            raise HTTPException(status_code=422, detail=str(exc)) from exc
        return {"status": "ok", "merge_id": merge_id}

    return router
